File: src/runtime/boxes/bb_brk.c
/* _XBRKC    BRK         scan to first char in set (may be zero-width) */
#include "bb_box.h"
#include <stdlib.h>
#include <string.h>


spec_t bb_brk(void *zeta, int entry)
{
    brk_t *ζ = zeta;
    spec_t BRK;
    if (entry==α)                                                               goto BRK_α;
    if (entry==β)                                                               goto BRK_β;
    BRK_α:          ζ->δ=0;                                                     
                    while (Δ+ζ->δ<Ω && !strchr(ζ->chars,Σ[Δ+ζ->δ])) ζ->δ++;     
                    if (Δ+ζ->δ >= Ω)                                            goto BRK_ω;
                    BRK = spec(Σ+Δ,ζ->δ); Δ += ζ->δ;                            goto BRK_γ;
    BRK_β:          Δ -= ζ->δ;                                                  goto BRK_ω;
    BRK_γ:                                                                      return BRK;
    BRK_ω:                                                                      return spec_empty;
}

brk_t *bb_brk_new(const char *chars)
{ brk_t *ζ=calloc(1,sizeof(brk_t)); ζ->chars=chars; return ζ; }

File: src/runtime/boxes/BbBrk.java
/**
 * BbBrk.java — BRK (BREAK): scan to first char in set (zero-width possible)
 * Port of bb_brk.c / bb_brk.s
 *
 *   BRK_α:  δ=0; while (Δ+δ<Ω && !chars.has(Σ[Δ+δ])) δ++;
 *           if (Δ+δ >= Ω)               goto BRK_ω;   (never found char in set)
 *           BRK=spec(Σ+Δ,δ); Δ+=δ;     goto BRK_γ;
 *   BRK_β:  Δ-=δ;                       goto BRK_ω;
 */
class BbBrk extends BbBox {
    private final String chars;
    private int          delta;

    public BbBrk(MatchState ms, String chars) { super(ms); this.chars=chars; }

    @Override public Spec alpha() {
        delta = 0;
        while (ms.delta + delta < ms.omega &&
               chars.indexOf(ms.sigma.charAt(ms.delta + delta)) < 0) delta++;
        if (ms.delta + delta >= ms.omega) return null;
        Spec r = new Spec(ms.delta, delta);
        ms.delta += delta;
        return r;
    }

    @Override public Spec beta() { ms.delta -= delta; return null; }
}
